vms/platformvm/validators/manager_test.go

```diff
@@ -0,0 +1,445 @@
+				{
+					vdrs[1].NodeID: vdrs[1].PublicKey,
+				},
+			},
+			expectedVdrSet: map[ids.NodeID]*validators.GetValidatorOutput{
+				vdrs[0].NodeID: {
+					NodeID:    vdrs[0].NodeID,
+					PublicKey: vdrs[0].PublicKey,
+					Weight:    vdrs[0].Weight,
+				},
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+
+			// setup validators set
+			vdrs := validators.NewMockManager(ctrl)
+			cfg := config.Config{
+				Chains:                 chains.TestManager,
+				UptimePercentage:       .2,
+				RewardConfig:           defaultRewardConfig,
+				Validators:             vdrs,
+				UptimeLockedCalculator: uptime.NewLockedCalculator(),
+				BanffTime:              mockable.MaxTime,
+			}
+			mockState := state.NewMockState(ctrl)
+
+			metrics, err := metrics.New("", prometheus.NewRegistry(), cfg.TrackedSubnets)
+			r.NoError(err)
+
+			clk := &mockable.Clock{}
+			validatorssSet := NewManager(cfg, mockState, metrics, clk)
+
+			// Mock the VM's validators
+			mockSubnetVdrSet := validators.NewMockSet(ctrl)
+			mockSubnetVdrSet.EXPECT().List().Return(tt.currentSubnetValidators).AnyTimes()
+			vdrs.EXPECT().Get(tt.subnetID).Return(mockSubnetVdrSet, true).AnyTimes()
+
+			mockPrimaryVdrSet := mockSubnetVdrSet
+			if tt.subnetID != constants.PrimaryNetworkID {
+				mockPrimaryVdrSet = validators.NewMockSet(ctrl)
+				vdrs.EXPECT().Get(constants.PrimaryNetworkID).Return(mockPrimaryVdrSet, true).AnyTimes()
+			}
+			for _, vdr := range tt.currentPrimaryNetworkValidators {
+				mockPrimaryVdrSet.EXPECT().Get(vdr.NodeID).Return(vdr, true).AnyTimes()
+			}
+
+			// Tell state what diffs to report
+			for _, weightDiff := range tt.weightDiffs {
+				mockState.EXPECT().GetValidatorWeightDiffs(gomock.Any(), gomock.Any()).Return(weightDiff, nil)
+			}
+
+			for _, pkDiff := range tt.pkDiffs {
+				mockState.EXPECT().GetValidatorPublicKeyDiffs(gomock.Any()).Return(pkDiff, nil)
+			}
+
+			// Tell state last accepted block to report
+			mockTip := blocks.NewMockBlock(ctrl)
+			mockTip.EXPECT().Height().Return(tt.lastAcceptedHeight)
+			mockTipID := ids.GenerateTestID()
+			mockState.EXPECT().GetLastAccepted().Return(mockTipID)
+			mockState.EXPECT().GetStatelessBlock(mockTipID).Return(mockTip, choices.Accepted, nil)
+
+			// Compute validator set at previous height
+			gotVdrSet, err := validatorssSet.GetValidatorSet(context.Background(), tt.height, tt.subnetID)
+			r.ErrorIs(err, tt.expectedErr)
+			if tt.expectedErr != nil {
+				return
+			}
+			r.Equal(len(tt.expectedVdrSet), len(gotVdrSet))
+			for nodeID, vdr := range tt.expectedVdrSet {
+				otherVdr, ok := gotVdrSet[nodeID]
+				r.True(ok)
+				r.Equal(vdr, otherVdr)
+			}
+		})
+	}
+}
+
+func copyPrimaryValidator(vdr *validators.Validator) *validators.Validator {
+	newVdr := *vdr
+	return &newVdr
+}
+
+func copySubnetValidator(vdr *validators.Validator) *validators.Validator {
+	newVdr := *vdr
+	newVdr.PublicKey = nil
```

Comment:
why do we set the public key for subnet validators to nil?

Reply:
my understanding is that primary keys are registered when primary network validator is created (via an AddPermissionlessValidatorTx)
Here we are trying to emulate the creation of a subnet validator, for which PubKey should not be set. 
I'd ask confirmation from @danlaine, who created the test in the first place.

Reply:
Subnet validators should not have BLS keys (BLS keys are shared by the Primary Network validator)